File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.swift
import Foundation

func solution(_ numbers:[Int], _ target:Int) -> Int {
    var result = 0
    var sum_arr = [0]   // 반복문 시작을 위해 0 하나 추가
    
    for n in numbers {
        var temp: [Int] = []
        for s in sum_arr {
            temp.append(s+n)
            temp.append(s-n)
        }
        sum_arr = temp
    }
    
    for s in sum_arr { result += (s == target) ? 1 : 0 }
    
    return result
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    static int cnt = 0;
    public int solution(int[] numbers, int target) {
        dfs(numbers, 0, 0, target);
        return cnt;
    }
    
    public void dfs(int[] numbers, int idx, int result, int target) {
        if (idx == numbers.length) {
            if (result == target) cnt++;
            return;
        } else {
            dfs(numbers, idx+1, result+numbers[idx], target);
            dfs(numbers, idx+1, result-numbers[idx], target);
        }
    }
}